Main (#1)
lumc-sasc/wf-rnaseq: docs/DEVELOPERGUIDE.md

# DEVELOPER GUIDE

> This is the developers guide for the RNA-seq workflow. The workflow is being developed with the purpose of geting information from the reads, including the gene expression.

Reviewer: ```suggestion
This is the developers guide for the RNA-seq workflow. The workflow is being developed with the purpose of getting information from the reads, including the gene expression.
```

The framework used to create this workflow is Nextflow.
The purpose of the guide is to give a bit of insight regarding working with the worfklow. This includes Architecture and dependencies.

# Table of Contents
1. [Layout](#layout)
   - [Workflow Image](#workflow-image)
   - [Directory structure](#directory-structure)
2. [Dependencies](#dependencies)
   - [Workflow dependencies](#workflow-dependencies)
3. [Workflow](#workflow)
   - [Workflow](#workflow)
   - [Subworkflow](#subworkflow)
   - [Process](#process)
   - [Configuration file](#configuration-file)
   - [Dynamic resource allocation](#dynamic-resource-allocation)
3. [Testing](#testing)
   - [nf-test](#nf-test)
   - [Pytest workflow](#pytest-workflow)
4. [Development](#development)
   - [Adding features](#adding-features)

# Layout

## Workflow Image
![Workflow Image](https://github.com/lumc-sasc/wf-rnaseq/blob/main/docs/images/Workflow_Diagram.png)

> The workflow image shows all the processes that exist in the current model of the pipeline. <br/>

Reviewer: `<br/>` is not needed. If you want to explicitly insert a linebreak, you can just put two spaces (`  `) at the end of a line  in markdown.

Each color describes a different part of the pipeline and each shape describes either a file or a process. <br/>
The rectangle with the wave is the mandatory file block. The mandatory file block is the block that shows the mandatory files. <br/>
A rectangle with two vertical lines is the mandatory process. Those processes have to be run in the pipeline. <br/>
The parallel square is the between steps that happen between processes. <br/>
The diamond shape is all the true/false and yes/no statements. <br/>
The hexagon is all the optional files and variables that are either given as input or generated as output. <br/>
The non edged square is for all the processes that are not mandatory and are optional. <br/> <br/>

> Fastqc subworkflow falls under the blue part of the workflow image. <br/>

Reviewer: A blank line should already be enough, no need for `<br/> <br/>`.

Sample subworkflow falls under the yellow part of the workflow image. <br/>
Bammetrics subworkflow falls under the red/pink part of the workflow image. <br/>

> All of variantcalling falls under the purple part of the workflow image. It is currently not finished yet. <br/>

Reviewer: > It is currently not finished yet.

Is this still true?

LncRNAseq falls under the dark green part of the workflow image. <br/>
Expression quantification subworkflow falls under the dark red and grey part of the workflow image <br/><br/><br/>

> <hr><br/><br/>

Reviewer: ```suggestion

---

```

## Directory structure.
A certain structure should be followed when addding processes, workflows and subworkflows. <br/>
A workflow should be added to the workflows directory, <br/>
a subworkflow should be added to the subworkflows directory, <br/>
and a process should be added in the local directory of the modules directory if it is a custom module. <br/>
The script that should be initiated is the workflow hub file. <br/>

> configurations should be in the config directory, with the exception of the nextflow.config file, which should be in the same directory as the main.nf script.

Reviewer: ```suggestion
Configurations should be in the config directory, with the exception of the nextflow.config file, which should be in the same directory as the main.nf script.
```

<hr><br/><br/>

# Dependencies
## Workflow dependencies

> For workflow dependencies, see user guide.

Reviewer: Please add a link to the user guide.

The following example shows how to write an hub: <br/>

```nextflow
//import workflows and reading modules
include { RNA_seq } from './workflows/RNA-seq'

//Runs the local nextfow RNA-seq pipeline developed by SASC
workflow RNA_seq_pipeline {
   RNA_seq()
}

//Runs the Nanoseq pipeline (example)
workflow Nanoseq_pipeline {
   Nanoseq()
}
```

How it is called on in the bash environment:
```bash
nextflow run main.nf -entry RNA_seq_pipeline args
```
<hr><br/><br/>

## Workflow
The main workflow runs the workflow itself. It calls all the subworkflows that has to be run, with the exception of nested subworkflows. <br/>
The following example shows how to write a main workflow: <br/>
```nextflow
//This part describes the input files.
Samplesheet = params.samplesheet
Fasta = params.fasta
FastaFai = params.fastafai

//Start of the workflow iself.
workflow RNA-seq {
   // A subworkflow gets called up.
   subworkflow1(Samplesheet, Fasta, FastaFai)

   // A process that uses the output of the subworkflow and a variable from the configuration file.
   process1(subworkflow1.out.reads, params.strandedness)
}
```
<hr><br/><br/>

## Subworkflow

> The subworkflow runs the processes of a certain part of the workflow. The subworkflows are used to catagorize everything. <br/>

Reviewer: ```suggestion
The subworkflow runs the processes of a certain part of the workflow. The subworkflows are used to categorize everything.
```

      memory = {1.B * index.target.size() * 1.3 * (task.attempt * 0.1 + 0.9)}

      //Allocation of time. It first looks if it is a simple file or multiple files.
      //The reason for this being is that if one attempts to go through a reduce with only a single file, it will produce an error.
      //If there are multiple files, it will first grab all the full paths of the files using the .target.
      //It will then go through each instance with the use of stream.
      //While it goes through each instance, it will grab the fil size and add up the size with the total value with the use of .reduce.
      //After that it multiplies by the attempt of the process run.
      time = { reads.toList().size() > 1 ?
         1.ms * ((reads.target.stream().reduce(0, (x, y) -> x + y.size())) * task.attempt / 1000 + 300000) :
         1.ms * (reads.target.size() * task.attempt / 1000 + 125000)}

   }
}

```

# Testing

## nf-test
nf-test is a testing framework developed by some of the developers of Nextflow. It allows for testing the working of the workflows and processes. One of the things it can see for example is whether an output file is present.

> Right now, there is an example of nf-test in the Htseq subdirectory within the test directory. It looks for the corrolation between the output file generated and the reference file. 

Reviewer: ```suggestion
Right now, there is an example of nf-test in the Htseq subdirectory within the test directory. It looks for the correlation between the output file generated and the reference file. 
```
Is this test looking at correlation? Shouldn't it be looking at concordance?

If the percentage is above 99%, it passes. Else, it will fail.

To be able to work with nf-test, nf-test dependency have to be installed alongside the ones for Nextflow. Follow the following code to install dependency:
```bash
conda activate nextflow
conda install -c bioconda nf-test
```
<br/>
nf-test file step by step walkthrough

```nextflow
nextflow_workflow {
}
//This part describes what type of testcase it is. In this case, it is ia workflow test so the test is wrapped in a nextflow_workflow test wrap.

> //Everything is desribed within these code brackets.

Reviewer: Why is this a comment within the code block? It would probably be easier to read if this is just a paragraph underneath or above the code.

```
<br/>

```nextflow
name "Test worfklow with Htseqcount"
script "../../main.nf"
workflow "RNA_seq"
//This part has the basic information, like what the name of the test will be, which file it should execute and what the name is of the workflow.
```
<br/>

```nextflow
stage {
   symlink "/exports/sascstudent/vperinbanathan"
}
//Staging of directory. This is needed, because otherwise it won't find the files neccesary, including inputfiles.
```
<br/>

```nextflow

> test("Corrolation of Collect column counts has to be higher than 99.9%") {
> }
> //Descrption of the test and the execution of the test. Both everything that happens within the workflow execution as what happens after will happen within these code brackets.

Reviewer: ```suggestion
test("Correlation of Collect column counts has to be higher than 99.9%") {
}
//Description of the test and the execution of the test. Both everything that happens within the workflow execution as what happens after will happen within these code brackets.
```

```bash
conda activate nextflow
conda install pytest-workflow
```

<br/>

test_RNA-seq.yml
```yml
#The name of the test. If one wants to run a specific test, it is based on the name that has been given to the test.
- name: Htseq_corrolation

#The exact linux command that it will execute.
  command: nextflow run (full filepath to workflow hub) -entry RNA_seq_pipeline --genome 'Nextflow_test_human' --sampleConfigFile ./test/data/samplesheet.yml --outdir ../test/outputfiles

  #These two lines will look if the expected outputfile has been generated. If it has not been generated, it will generate an error.
  files:
    - path: "../test/outputfiles/final_gene_count/report/report_collect_0.csv"

```

<br/>

test_corrolation.py
```python
#This part of the script includes the packages needed and creates the custom test based on the test name. In this case, because the test in the yml file is Htseq_corrolation, it should be in here too.
#It will run the function that is below the creation of the test. AFter declaration of function, code can be executed to 
import pathlib
import pytest
@pytest.mark.workflow('Htseq corrolation')
def test_corrolation_check():
```

<br/>

Code inside function

> ```python

Reviewer: Why separate this into multiple code blocks?

        So it is very difficult to check which of the two should be removed if there is a form of discrepancy.
        """
        else:
            del count_lines_reference[Reference_lines]
            Reference_lines -= 1
            del count_lines_nextflow[Nextflow_lines]
            Nextflow_lines -= 1

    #Calculates the total corrolation in percentage
    output_corrolation = corrolation/line_count * 100

    #Checks whether the corrolation is higher than 99%. If the corrolation is lower than 99%, it will generate an error message.
    assert output_corrolation > 99.0

```

To be able to work with pytest workflow, a tutorial can be followed on the following link: [Link to tutorial](https://pytest-workflow.readthedocs.io/en/stable/)

# Development

## Adding features
When wanting to add a feature to the existing workflow and want to apply it globally, please open an issue to discuss what you would like to change. For own use cases, adding features to local environment poses no issues.
Please do write code documentation when adding a new feature and update when needed.

> 
> 
> 
> 
> 
>                                                                                                                                                                                                                                                                                                                                                                               
> 
> 
> 
> 

Reviewer: ```suggestion                                                                                                                                                                                                                                                                                                                                                                   
```
